src/lib/include/x86_swizzle.h

```diff
@@ -2,97 +2,226 @@
 #define DISTOPIA_SWIZZLE_H
 
 #include "arch_config.h"
+#include "compiler_hints.h"
+#include "distopia_type_traits.h"
 
 #ifdef DISTOPIA_X86_SSE4_1
 
-#include <immintrin.h>
-
 #include "x86_tgintrin.h"
+#include <emmintrin.h>
+#include <immintrin.h>
 
 namespace {
 
-inline void Deinterleave3(__m128 a, __m128 b, __m128 c,
-                          __m128& x, __m128& y, __m128& z) {
-  // a = x0y0z0x1, b = y1z1x2y2, c = z2x3y3z3
-  __m128 t1 = shuffle_p<_MM_SHUFFLE(2,1,3,2)>(b, c);
-  __m128 t2 = shuffle_p<_MM_SHUFFLE(1,0,2,1)>(a, b);
+// shuffle first element of __m128 to end and shunt everything left by one
+inline __m128 ShuntFirst2Last(const __m128 input) {
+  // PRE: input = abcd
+  return shuffle_p<_MM_SHUFFLE(0, 3, 2, 1)>(input, input);
+  // return bcda
+}
+
+#ifdef DISTOPIA_X86_AVX2_FMA
+
+// shuffle first element of __m256d to end and shunt everything left by one
+// NOTE requires AVX2 rather than AVX
```

Comment:
We should have an AVX1 version of this function as well so we don't require AVX2 for `__m256d`